ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl.hs

```diff
@@ -285,37 +285,59 @@ putBlockImpl env@VolatileDBEnv{ maxBlocksPerFile, tracer }
           , currentMaxSlotNo   = currentMaxSlotNo `max` MaxSlotNo bslot
           }
 
--- | The approach we follow here is to try to garbage collect each file.
--- For each file we update the fs and then we update the Internal State.
--- If some fs update fails, we are left with an empty Internal State and a
--- subset of the deleted files in fs. Any unexpected failure (power loss,
--- other exceptions) has the same results, since the Internal State will
--- be empty on re-opening. This is ok only if any fs updates leave the fs
--- in a consistent state every moment.
+-- NOTE: the current file is never garbage collected.
+
+-- | Garbage collect all files of which the highest slot is less than the
+-- given slot.
 --
--- This approach works since we always close the Database in case of errors,
--- but we should rethink it if this changes in the future.
+-- We first check whether we actually can garbage collect any file. If we can,
+-- we obtain the more expensive write lock and remove the files that can be
+-- garbage collected. We update the 'InternalState' for each garbage collected
+-- file.
+--
+-- If an exception is thrown while garbage collecting, we close the database.
+-- This means we don't have to worry the file system getting out of sync with
+-- the in-memory indices, as the indices are rebuilt when reopening.
 garbageCollectImpl :: forall m blockId. (IOLike m, Ord blockId)
                    => VolatileDBEnv m blockId
                    -> SlotNo
                    -> m ()
-garbageCollectImpl env slot =
-    writeOpenState env $ \hasFS -> do
-      files <- gets (sortOn fst . Index.toList . currentMap)
-      mapM_ (tryCollectFile hasFS slot) files
-      -- Recompute the 'MaxSlotNo' based on the files left in the VolatileDB.
-      -- This value can never go down, except to 'NoMaxSlotNo' (when we GC
-      -- everything), because a GC can only delete blocks < a slot.
-      modify $ \st -> st {
-          currentMaxSlotNo = FileInfo.maxSlotInFiles
-            (Index.elems (currentMap st))
-        }
+garbageCollectImpl env slot = do
+    -- Check if we can actually GC something using a cheaper read (allowing
+    -- for more concurrency) before obtaining the more expensive exclusive
+    -- write lock.
+    usefulGC <- atomically $ getterSTM gcPossible env
```

Comment:
Say that this is monotonic: if this is True, then `filesToGC` must be non-empty. That's because we're the only one GC'ing and because we never GC the current file.

No matter what happens in the middle, we only obtain the write lock when there is something to do.